dashboard/dashboard.go

```diff
@@ -0,0 +1,340 @@
+	Memory    []*data `json:"memory,omitempty"`
+}
+
+type data struct {
+	Time  time.Time `json:"time,omitempty"`
+	Value float64   `json:"value,omitempty"`
+}
+
+type status struct {
+	Peers int `json:"peers,omitempty"`
+	Block int `json:"block,omitempty"`
+}
+
+// New creates a new dashboard instance with the given configuration
+func New(config *Config) (*dashboard, error) {
+	//log.Trace("NewDashboard() called")
+
+	return &dashboard{
+		config:  config,
+		Metrics: &metricSamples{},
+		quit:    make(chan struct{}),
+	}, nil
+}
+
+// Protocols is a meaningless implementation of node.Service
+func (db *dashboard) Protocols() []p2p.Protocol { return nil }
+
+// APIs is a meaningless implementation of node.Service
+func (db *dashboard) APIs() []rpc.API { return nil }
+
+// Start implements node.Service, starting the data collection thread and the listening server of the dashboard
+func (db *dashboard) Start(server *p2p.Server) error {
+	//log.Trace("Start() called", "config", db.config)
+
+	go db.collectData()
+
+	http.HandleFunc("/", db.webHandler)
+	http.Handle("/api", websocket.Handler(db.apiHandler))
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", db.config.Host, db.config.Port))
+	if err != nil {
+		return err
+	}
+	db.listener = listener
+
+	go func() {
+		//log.Trace("Starting server...")
+
+		if err := http.Serve(listener, nil); err != nil {
+			log.Warn("Server failed", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop implements node.Service, stopping the data collection thread and the connection listener of the dashboard
+func (db *dashboard) Stop() error {
+	//log.Trace("Terminating dashboard...")
+
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	var err error
+	// Close the connection listener
+	if err = db.listener.Close(); err != nil {
+		log.Warn("Failed to close listener", "err", err)
+	}
+
+	close(db.quit) // Notifies collectData and apiHandler
+
+	for _, c := range db.conns {
+		if err := c.conn.Close(); err != nil {
+			c.logger.Warn("Failed to close connection", "err", err)
+		}
+	}
+	db.conns = db.conns[:0]
+
+	return err
+}
+
+// webHandler handles all non-api requests, simply flattening and returning the dashboard website.
+func (db *dashboard) webHandler(w http.ResponseWriter, r *http.Request) {
+	//log.Trace("webHandler() called", "r.URL", r.URL)
+	log.Info("Request", "URL", r.URL)
+
+	path := r.URL.String()
+
+	// If the path of the assets is manually set
+	if db.config.Assets != "" {
+		// Create the filename for ReadFile
+		var buffer bytes.Buffer
+		buffer.WriteString(db.config.Assets)
+		switch path {
+		case "/":
+			buffer.WriteString("/dashboard.html")
+		default:
+			buffer.WriteString(path)
+		}
+
+		file, err := ioutil.ReadFile(buffer.String())
+		if err != nil {
+			// TODO (kurkomisi): Warn or Crit?
+			log.Warn("Failed to read file", "err", err)
+			// TODO (kurkomisi): Should I inform the client?
+			return
+		}
+		w.Write(file)
+		return
+	}
+
+	switch path {
```

Comment:
Why do you need this switch on the path? Simply load `Asset(path)` and serve it to the user.